### DISCUSSION TITLE
Helix Generic Content Model

### DISCUSSION THREAD

**Discussion by tripodsan** [adobe/helix-html2md#6] [Ideas]
> Work in progress document to describe the helix generic content model in a formal way.

Semantically, a Helix Generic Content document follows the [HTML5][htmlContent] content model, with some limitations.

Note that the document generated by the rendering pipeline can contain more elements than the ones that describe the semantic content. eg. the pipeline includes the `head.html` verbatim, which can contain any [metadata][metadata] elements or gibberish.

## outline

Since the overall structure is a [HTML5][htmlContent] document, it is not described here too formally, but given in exemplary form:

```html
<!DOCTYPE html>
<html>
  <head>
    {metadata}
  </head>
  <body>
    <header>{headerContent}</header>
    <main>{mainContent}</main>
    <footer>{footerContent}</footer>
  </body>
</html>
```

```idl
interface HelixDocument {
  children: [ Head | Body ];
}

interface Head {
  children: [ HelixMetadata ];
}

interface Body {
  children: [ HeaderContent | MainContent | FooterContent ];
}
```

```ebnf
helix := "<!DOCTYPE html>" html;
html := "<html>" head body "</html>";
head := "<head>" metadata* "</head>";
metadata := "<meta name=... content=... >" | title;
title := "<title>" text "</title>";
body := "<body>" header main footer "</body>";
header := "<header />";
footer := "<footer />";
```
### Metadata content

Helix recognises the `<meta>` and `<title>` [metadata][metadata] elements. All others are ignored on import, but can be present on export.

```idl
interface HelixMetadata {
  children: [ Meta* | Title ];
}

interface Meta {
  name: string;
  content: string;
}

interface Title {
  children: [ Text ];
}
```

### Header content

Always empty (TBD)

### Footer content

Always empty (TBD)

### Main Content

All the content that semantically maps to document semantics of an individual page or fragment is found in the [main-content][mainContent] content section of the markup. It consists of default content in sections and blocks.

```ebnf
main := "<main>" section* "</main>";
section := "<div>" sectionContent "</div>";
```

```idl
interface MainContent {
  children: [ Section* ];
}

interface Section {
  children: [ FlowContent | Block ]*;
}
```

The following html elements, along with their respective semantics are understood; those are a subset of the [flow content][flowContent] defined by html.

```ebnf
helixFlowContent :=
  | "<a>"
  | "<br>"
  | "<code>"
  | "<del>"
  | "<em>"
  | "<h1>" | "<h2>" | "<h3>" | "<h3>" | "<h4>" | "<h5>" | "<h6>"
  | "<img>"
  | "<li>"
  | "<ol>"
  | "<p>"
  | "<picture>"
  | "<pre>"
  | "<source>"
  | "<span>"
  | "<strong>"
  | "<sub>"
  | "<sup>"
  | "<table>"
  | "<tbody>"
  | "<td>"
  | "<th>"
  | "<thead>"
  | "<tr>"
  | "<u>"
  | "<ul>";
```

[metadata]: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#metadata_content
[htmlContent]: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#main_content_categories
[flowContent]: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#flow_content
[mainContent]: #main-content

**Comment by tripodsan** [comment]
### Section Content

Section content can contain phrasing content and block. The flow content is understood in its semantic meaning. For example a `<h1>` is a title, a `<em>` marks text as emphasised (and not per se italic).

```ebnf
sectionContent := ( helixFlowContent | block )*;
```

### Blocks

A helix _block_ is identified by nested `<div>` tags which can be used to further structure the document by adding attributes to parts of the content. The content of a block cell is again semantic html, but please note that nested blocks are not supported.

The `class` attribute of a block's `<div>` tag has special semantic, as it defines the block name and options.

```html
<div class=”blockname”>
  <div>
     <div>
      <p>Hello, World.</p>
     </div>
  </div>
</div>
```

```ebnf
block := "<div class=" blockClass ">" blockRow+ "</div>";
blockClass := blockName blockOptions*;
blockRow := "<div>" blockCell+ "</div>";
blockCell := "<div>" helixFlowContent* "</div>";
```

```idl
interface Block {
  name: string;
  options: string[];
  children: [ BlockRow ]*;
}

interface BlockRow {
  children: [ BlockCell ]*;
}

interface BlockCell {
  children: [ FlowContent ]*;
}
```

**Comment by bdelacretaz** [comment]
FWIW, if at some point you need more semantics in the HTML, note that we've been using [HTML Microdata](https://developer.mozilla.org/en-US/docs/Web/HTML/Microdata) for the Universal Editor, and that works well.

**Reply by tripodsan** [reply]
I would see this as an additional layer, where one could define some schema, that would decorate the microdata on the elements, based on block types etc. this should then also help transforming the updates back into the document.
